src/CaptureServiceBase/UploaderClientCaptureEventCollectorBuilder.cpp

```diff
@@ -6,21 +6,28 @@
 
 #include "ProducerEventProcessor/UploaderClientCaptureEventCollector.h"
 
+using orbit_capture_uploader::UploadDataInterface;
+using orbit_producer_event_processor::ClientCaptureEventCollector;
+using orbit_producer_event_processor::UploaderClientCaptureEventCollector;
+
 namespace orbit_capture_service_base {
 
-// A `ClientCaptureEventCollectorBuilder` implementation to build
-// `UploaderClientCaptureEventCollector` for the cloud collector.
-class UploaderClientCaptureEventCollectorBuilder : public ClientCaptureEventCollectorBuilder {
- public:
-  [[nodiscard]] std::unique_ptr<orbit_producer_event_processor::ClientCaptureEventCollector>
-  BuildClientCaptureEventCollector() override {
-    return std::make_unique<orbit_producer_event_processor::UploaderClientCaptureEventCollector>();
-  }
-};
-
-std::unique_ptr<ClientCaptureEventCollectorBuilder>
-CreateUploaderClientCaptureEventCollectorBuilder() {
-  return std::make_unique<UploaderClientCaptureEventCollectorBuilder>();
+std::unique_ptr<ClientCaptureEventCollector>
+UploaderClientCaptureEventCollectorBuilder::BuildClientCaptureEventCollector() {
+  std::unique_ptr<UploaderClientCaptureEventCollector> uploader_client_capture_event_collector =
+      std::make_unique<UploaderClientCaptureEventCollector>();
+
+  absl::MutexLock lock{&mutex_};
+  upload_data_interface_ = uploader_client_capture_event_collector.get();
```

Comment:
Hmmm. This looks like it could potentially be a problem for us?

When does `uploader_client_capture_event_collector` get destroyed? I assume it will be destroyed by the `CaptureServiceBase` class, right?

When that happens, `upload_data_interface_` will become a dangling pointer. Can this be a problem for the user of this pointer?

Reply:
If so, maybe we should replace the return type of `BuildClientCaptureEventCollector` to a raw pointer and let the builder manage the lifetime of the collector. Just a thought...

Reply:
Cool. But are you aware that this means we also need to change the other `ClientCaptureEventCollectorBuilder` for OrbitService? Not sure how much work that is.

Reply:
I made the following change:
* rename `ClientCaptureEventCollectorBuilder` as `ClientCaptureEventCollectorManager`, and it provides only the `GetClientCaptureEventCollector` method;
* let `ClientCaptureEventCollectorManager` own the created `ClientCaptureEventCollector`. The manager will create the collector in the constructor and provides a raw pointer to the collector in `GetClientCaptureEventCollector` method;
* change the `CaptureServiceBase` to own the `ClientCaptureEventCollectorManager` rather than the `ClientCaptureEventCollector`